docs/task_manager_system.md

```diff
@@ -1,33 +1,113 @@
 1. Creation time
 2. Job dependencies
 3. Capacity
 
-Independent jobs are run in order of creation time, earliest first. Jobs with dependencies are also run in creation time order within the group of job dependencies. Capacity is the final consideration when deciding to release a job to be run by the task dispatcher.
+Independent tasks are run in order of creation time, earliest first. Tasks with dependencies are also run in creation time order within the group of task dependencies. Capacity is the final consideration when deciding to release a task to be run by the dispatcher.
+
+
+## Dependency Manager
+
+Responsible for looking at each pending task, and determining whether it should create a dependency for that task.
+For example, if `update_on_launch` is enabled of a task, a project update will be created as a dependency of that task. The Dependency Manager is responsible for creating that project update.
+
+Dependencies can also have their own dependencies, for example,
+
+```
++-----------+
+|           |                          created by web API call
+|   Job A   |
+|           |
++-----------+---+
+                |
+                |
+        +-------v----+
+        |  Inventory |                 dependency of Job A
+        |  Source    |                 created by Dependency Manager
+        |  Update B  |
+        +------------+-------+
+                             |
+                             |
+                      +------v------+
+                      |   Project   |  dependency of Inventory Source Update B
+                      |   Update C  |  created by Dependency Manager
+                      +-------------+
+```
+
+
+### Dependency Manager Algorithm
+
+1. Get pending tasks (parent tasks) that have `dependencies_processed = False`
+2. Create project update if
+    a. not already created
+    b. last project update outside of cache timeout window
+3. Create inventory source update if
+    a. not already created
+    b. last inventory source update outside of cache timeout window
+4. Check and create dependencies for these newly created dependencies
+    a. inventory source updates can have a project update dependency
+5. All dependencies are linked to the parent task via the `dependent_jobs` field
+    a. This allows us to cancel the parent task if the dependency fails or is canceled
+6. Update the parent tasks with `dependencies_processed = True`
+
+
+## Task Manager
+
+Responsible for looking at each pending task and determining whether it can start that task.
+
 
-## Task Manager Architecture
+### Task Manager Algorithm
+
+1. Get pending, waiting, and running tasks that have `dependencies_processed = True`
+2. For each running task, add task to dependency graph
+    a. dependency graph is just an internal data structure that tracks which
+    jobs are currently running. It also handles "soft" blocking logic
```

Comment:
in same loop that we add tasks to dependency graph, we do the initial calculation of currently consumed capacity

```
 Before processing pending tasks, the task manager first processes running tasks. This allows it to build a dependency graph and account for the currently consumed capacity in the system.
    a. dependency graph is just an internal data structure that tracks which
    jobs are currently running. It also handles "soft" blocking logic
    b. the capacity is tracked in memory on the TaskManagerInstances and TaskManagerInstanceGroups objects which are  in-memory representations of the instances and instance groups. These data structures are used to help track what consumed capacity will be as we decide that we will start new tasks, and until such time that we actually commit the state changes to the database.
```

Reply:
added, thanks